### PR TITLE
Add libServer and API to lookup nodes

### PR DESCRIPTION
**Pull request by sidthekidder**

### CODE REVIEW

**Comment by sidthekidder** [comment]
@sharwell Libjson-rpc-cpp required a slight modification to run on ubuntu 14 (https://gist.github.com/stefanofiorentino/594858159c23841ec4eec56184852d51) so I pointed it to my own fork, and am building it from source. Is this sustainable or should we look for some other solution?

**Comment by sharwell** [comment]
@sidthekidder I'm trying to find the Travis build where the original code was failing. It looks like the commit history from #79 is gone. Do you know which build it was?

**Comment by sidthekidder** [comment]
@sharwell this is one of them https://travis-ci.org/Zilliqa/Zilliqa/jobs/345004085

**Comment by sharwell** [comment]
@sidthekidder Thanks. It looks like the issue occurs if you don't have libmicrohttpd 0.9.39 or newer, per the following release notes:

https://github.com/svn2github/libmicrohttpd/blob/e4a96ac6b99bc120b34c6af4ed9520d8b50c355e/ChangeLog#L522-L531

It looks like 0.9.33-1 is installed on these machines by default. An approach like this, if it works, would be preferable:

https://github.com/dupgit/sauvegarde/blob/fa5c0156aee008ec4ecd6b21a9c05d150652e7b8/.travis.yml#L28-L30

**Comment by sharwell** [comment]
Looks like you may need to tweak the argument `--prefix=$HOME/local` for the MHD build, and possibly sudo for installing it. I can take a closer look tomorrow.

**Comment by sidthekidder** [comment]
@sharwell Yup, that worked nicely, thanks

File: .travis.yml [hunk 1]
```diff
@@ -38,3 +41,10 @@ script:
 matrix:
   allow_failures:
     - os: osx
+
+install:
+  # libmicrohttpd (the one in ubuntu 12.04 is too old)
```
Comment (bb111189) [inline .travis.yml]: Travis is using 14.04. Is it possible to get the packages from repo directly?

Comment (sidthekidder) [reply .travis.yml]: The default libmicrohttpd isn't updated https://travis-ci.org/Zilliqa/Zilliqa/jobs/346214265 , so explicitly fetching this version

**Comment by bb111189** [comment]
In file included from /home/junhao/Desktop/octcoin/other_br/codereview/libserver/Zilliqa/src/libServer/Server.cpp:20:0:
/usr/include/jsonrpccpp/server/connectors/httpserver.h:29:24: fatal error: microhttpd.h: No such file or directory

Compilation error.

We need to update readme to reflect all the necessary dependencies before I can accept this PR.

**Comment by sidthekidder** [comment]
@bb111189 have added steps, let me know if any error still occurs

File: src/libServer/Server.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+#include "common/Serializable.h"
+
+#include "Server.h"
+
+using namespace jsonrpc;
+using namespace std;
+using namespace boost::multiprecision;
+
+
+Server::Server() : AbstractZServer(*(new HttpServer(4201)))
```
[the remark this answers is not in the document]
Comment (sidthekidder) [reply src/libServer/Server.cpp]: Will fix in issue #96

**Comment by bb111189** [comment]
@Gnnng Do help me test it on Mac

File: .travis.yml [hunk 3]
```diff
@@ -38,3 +41,9 @@ script:
 matrix:
   allow_failures:
     - os: osx
+
+install:
+  - wget -c http://ftp.gnu.org/gnu/libmicrohttpd/libmicrohttpd-0.9.59.tar.gz
+  - (tar zxf libmicrohttpd-0.9.59.tar.gz; cd libmicrohttpd-0.9.59; ./configure; make; sudo make install; cd ..; rm -fr libmicrohttpd-0.9.59)
+  - wget -O libjson-rpc-cpp-1.1.0.tar.gz https://github.com/cinemast/libjson-rpc-cpp/archive/v1.1.0.tar.gz
+  - (tar xzf libjson-rpc-cpp-1.1.0.tar.gz; cd libjson-rpc-cpp-1.1.0; cmake . && make; sudo make install; sudo ldconfig; cd ..; rm -fr libjson-rpc-cpp-1.1.0)
```
Comment (Gnnng) [inline .travis.yml]: #100

File: README.md [hunk 4]
```diff
@@ -43,16 +43,31 @@ To compile and run the Zilliqa codebase, you will need the following dependencie
...
+sudo apt-get install curl libargtable2-dev libhiredis-dev
+wget -c http://ftp.gnu.org/gnu/libmicrohttpd/libmicrohttpd-0.9.59.tar.gz
+(tar zxf libmicrohttpd-0.9.59.tar.gz; cd libmicrohttpd-0.9.59; ./configure; make; sudo make install; cd ..; rm -fr libmicrohttpd-0.9.59)
+wget -O libjson-rpc-cpp-1.1.0.tar.gz https://github.com/cinemast/libjson-rpc-cpp/archive/v1.1.0.tar.gz
+(tar xzf libjson-rpc-cpp-1.1.0.tar.gz; cd libjson-rpc-cpp-1.1.0; cmake . && make; sudo make install; sudo ldconfig; cd ..; rm -fr libjson-rpc-cpp-1.1.0)
+```
+
+For higher _Ubuntu_ versions, you may install libjson-rpc-cpp and libmicrohttpd directly:
+```bash
+sudo apt-get install libjsonrpccpp-dev libmicrohttpd-dev
```
Comment (Gnnng) [inline README.md]: #100

**Comment by Gnnng** [comment]
@sidthekidder #100 enables using Ubuntu 16.04 on travis, can you change `.travis.yml` and `README.md` on your branch a bit? We do not intend to maintain the dependencies for other distros currently, so we can remove the related instructions for now.

Another nice thing to have is we can rely on the official versions of `libmicrohttpd` and `libjsonrpccpp` and not manually compile and install them anymore. Can you confirm these changes as well?

**Comment by sidthekidder** [comment]
@Gnnng yes, the min version simplifies the dependencies, thanks. I've made the changes though I see that explicitly installing libmicrohttpd-dev is still required before libjsonrpccpp-dev (otherwise it fails)

File: README.md [hunk 5]
```diff
@@ -47,15 +47,16 @@ To compile and run the Zilliqa codebase, you will need the following dependencie
 * `CMake`
 * `build-essential`
 * `pkg-config` 
+* `json-rpc-cpp`
 
 For Ubuntu 16.04, you can use the following command to install the dependencies:  
 ```bash
-sudo apt-get install libboost-all-dev libssl-dev libleveldb-dev libjsoncpp-dev libsnappy-dev cmake build-essential pkg-config
+sudo apt-get install libboost-all-dev libssl-dev libleveldb-dev libjsoncpp-dev libsnappy-dev cmake libjsonrpccpp-dev build-essential pkg-config
```
Comment (Gnnng) [inline README.md]: @sidthekidder just a minor question here: don't we need `libmicrohttpd-dev` here as well? `libmicrohttpd-dev` seems not existing on a fresh Ubuntu 16.04.

Comment (sidthekidder) [reply README.md]: updated 👍

**Pull request merged**